src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipeEventDispatcher.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Diagnostics.Tracing
+{
+#if FEATURE_PERFTRACING
+    internal sealed class EventPipeEventDispatcher
+    {
+        internal static EventPipeEventDispatcher Instance = new EventPipeEventDispatcher();
+
+        private IntPtr m_RuntimeProviderID;
+
+        private bool m_stopDispatchTask;
+        private Task m_dispatchTask = null;
+        private object m_dispatchTaskLock = new object();
+
+        // For now, only one EventListener at a time gets to control what events are enabled.
+        private EventListener m_controllingEventListener = null;
+
+        private EventPipeEventDispatcher()
+        {
+            // Get the ID of the runtime provider so that it can be used as a filter when processing events.
+            m_RuntimeProviderID = EventPipeInternal.GetProvider(RuntimeEventSource.EventSourceName);
+        }
+
+        internal void SendCommand(EventListener eventListener, EventCommand command, bool enable, EventLevel level, EventKeywords matchAnyKeywords)
+        {
+            if (command == EventCommand.Update && enable)
+            {
+                EventPipeProviderConfiguration[] providerConfiguration = new EventPipeProviderConfiguration[]
+                {
+                    new EventPipeProviderConfiguration(RuntimeEventSource.EventSourceName, (ulong) matchAnyKeywords, (uint) level)
+                };
+
+                lock (m_dispatchTaskLock)
+                {
+                    if (m_dispatchTask == null)
+                    {
+                        m_controllingEventListener = eventListener;
+                        EventPipeInternal.Enable(null, 1024, 1, providerConfiguration, 1);
+                        m_dispatchTask = Task.Run(DispatchEventsToEventListeners);
+                    }
+                }
+            }
+            else if (command == EventCommand.Update && !enable)
+            {
+                RemoveEventListener(eventListener);
+            }
+        }
+
+        internal void RemoveEventListener(EventListener listener)
+        {
+            lock (m_dispatchTaskLock)
+            {
+                if (m_controllingEventListener == listener && m_dispatchTask != null)
+                {
+                    m_stopDispatchTask = true;
+                    m_dispatchTask.Wait();
```

Comment:
Good question.  We could consider a timeout here, but I'm not sure what it should be.  There is a question that I hadn't considered - what should the contract be between the disable operation and the dispatching of events?  Specifically, should the call to ```EventListener.DisableEvents``` return before or after we stop dispatching the events?  I suspect that the existing behavior is that we make no guarantees, in which case it would be safe to specify a timeout here.  The key though is that we don't want more than one dispatch task reading off the circular buffer, so we'd need to make sure that we can handle this appropriately even with a timeout.  For now, I'd say let's leave it as is, but I'm open to implementing this differently to ensure that we don't block here for longer periods of time than desired.